File: stacks/min_stack.py
"""LeetCode 155: Min Stack."""
import unittest


class MinStack:
    def __init__(self):
        self.values = []
        self.minimums = []

    def push(self, value):
        self.values.append(value)
        current_minimum = min(value, self.minimums[-1]) if self.minimums else value
        self.minimums.append(current_minimum)

    def pop(self):
        self.minimums.pop()
        return self.values.pop()

    def top(self):
        return self.values[-1]

    def get_min(self):
        return self.minimums[-1]


class MinStackTests(unittest.TestCase):
    def test_typical_case(self):
        stack = MinStack()
        for value in [-2, 0, -3]:
            stack.push(value)
        self.assertEqual(stack.get_min(), -3)
        stack.pop()
        self.assertEqual(stack.top(), 0)
        self.assertEqual(stack.get_min(), -2)

    def test_edge_case_duplicate_minimum(self):
        stack = MinStack()
        stack.push(2)
        stack.push(2)
        stack.pop()
        self.assertEqual(stack.get_min(), 2)


if __name__ == "__main__":
    unittest.main()

File: stacks/min_stack.c
#include <assert.h>
#include <stdio.h>

#define MAX_SIZE 10000

typedef struct {
    int values[MAX_SIZE];
    int minimums[MAX_SIZE];
    int size;
} MinStack;

void push(MinStack *stack, int value) {
    stack->values[stack->size] = value;
    stack->minimums[stack->size] = stack->size == 0 || value < stack->minimums[stack->size - 1]
        ? value : stack->minimums[stack->size - 1];
    stack->size++;
}

int pop(MinStack *stack) {
    return stack->values[--stack->size];
}

int top(const MinStack *stack) {
    return stack->values[stack->size - 1];
}

int get_min(const MinStack *stack) {
    return stack->minimums[stack->size - 1];
}

int main(void) {
    MinStack stack = {0};
    push(&stack, -2);
    push(&stack, 0);
    push(&stack, -3);
    assert(get_min(&stack) == -3);
    pop(&stack);
    assert(top(&stack) == 0 && get_min(&stack) == -2);

    MinStack duplicate_minimum = {0};
    push(&duplicate_minimum, 2);
    push(&duplicate_minimum, 2);
    pop(&duplicate_minimum);
    assert(get_min(&duplicate_minimum) == 2);
    puts("Min Stack: 2 tests passed");
    return 0;
}
